// couchbase/dist/binding.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CppTxnExternalException = exports.CppTxnFailureType = exports.CppPingState = exports.CppEndpointState = exports.CppSearchHighlightStyle = exports.CppSearchScanConsistency = exports.CppAnalyticsScanConsistency = exports.CppQueryProfileMode = exports.CppQueryScanConsistency = exports.CppViewSortOrder = exports.CppViewNamespace = exports.CppViewScanConsistency = exports.CppRetryReason = exports.CppErrorCode = exports.CppStatus = exports.CppSubdocOpcode = exports.CppSubdocStoreSemanticsType = exports.CppDurabilityLevel = exports.CppServiceType = void 0;
/* eslint jsdoc/require-jsdoc: off */
const bindings_1 = __importDefault(require("bindings"));
// enums
var CppServiceType;
(function (CppServiceType) {
})(CppServiceType = exports.CppServiceType || (exports.CppServiceType = {}));
var CppDurabilityLevel;
(function (CppDurabilityLevel) {
})(CppDurabilityLevel = exports.CppDurabilityLevel || (exports.CppDurabilityLevel = {}));
var CppSubdocStoreSemanticsType;
(function (CppSubdocStoreSemanticsType) {
})(CppSubdocStoreSemanticsType = exports.CppSubdocStoreSemanticsType || (exports.CppSubdocStoreSemanticsType = {}));
var CppSubdocOpcode;
(function (CppSubdocOpcode) {
})(CppSubdocOpcode = exports.CppSubdocOpcode || (exports.CppSubdocOpcode = {}));
var CppStatus;
(function (CppStatus) {
})(CppStatus = exports.CppStatus || (exports.CppStatus = {}));
var CppErrorCode;
(function (CppErrorCode) {
})(CppErrorCode = exports.CppErrorCode || (exports.CppErrorCode = {}));
var CppRetryReason;
(function (CppRetryReason) {
})(CppRetryReason = exports.CppRetryReason || (exports.CppRetryReason = {}));
var CppViewScanConsistency;
(function (CppViewScanConsistency) {
})(CppViewScanConsistency = exports.CppViewScanConsistency || (exports.CppViewScanConsistency = {}));
var CppViewNamespace;
(function (CppViewNamespace) {
})(CppViewNamespace = exports.CppViewNamespace || (exports.CppViewNamespace = {}));
var CppViewSortOrder;
(function (CppViewSortOrder) {
})(CppViewSortOrder = exports.CppViewSortOrder || (exports.CppViewSortOrder = {}));
var CppQueryScanConsistency;
(function (CppQueryScanConsistency) {
})(CppQueryScanConsistency = exports.CppQueryScanConsistency || (exports.CppQueryScanConsistency = {}));
var CppQueryProfileMode;
(function (CppQueryProfileMode) {
})(CppQueryProfileMode = exports.CppQueryProfileMode || (exports.CppQueryProfileMode = {}));
var CppAnalyticsScanConsistency;
(function (CppAnalyticsScanConsistency) {
})(CppAnalyticsScanConsistency = exports.CppAnalyticsScanConsistency || (exports.CppAnalyticsScanConsistency = {}));
var CppSearchScanConsistency;
(function (CppSearchScanConsistency) {
})(CppSearchScanConsistency = exports.CppSearchScanConsistency || (exports.CppSearchScanConsistency = {}));
var CppSearchHighlightStyle;
(function (CppSearchHighlightStyle) {
})(CppSearchHighlightStyle = exports.CppSearchHighlightStyle || (exports.CppSearchHighlightStyle = {}));
var CppEndpointState;
(function (CppEndpointState) {
})(CppEndpointState = exports.CppEndpointState || (exports.CppEndpointState = {}));
var CppPingState;
(function (CppPingState) {
})(CppPingState = exports.CppPingState || (exports.CppPingState = {}));
var CppTxnFailureType;
(function (CppTxnFailureType) {
})(CppTxnFailureType = exports.CppTxnFailureType || (exports.CppTxnFailureType = {}));
var CppTxnExternalException;
(function (CppTxnExternalException) {
})(CppTxnExternalException = exports.CppTxnExternalException || (exports.CppTxnExternalException = {}));
// Load it with require
const binding = (0, bindings_1.default)('couchbase_impl');
exports.default = binding;
